LangChain is an open-source framework uniquely designed to empower the development of applications leveraging large language models (LLMs). It stands out by providing essential tools and abstractions that enhance the customization, accuracy, and relevance of the information generated by these models.

At its core, LangChain offers a generic interface compatible with nearly any LLM. This facilitates a centralized development environment where data scientists can seamlessly integrate LLM applications with various external data sources and software workflows. This integration is crucial for those looking to harness the full potential of AI in their processes.

One of the most powerful features of LangChain is its module-based approach. This approach allows flexibility in performing experiments and optimizations of interactions with LLMs. Data scientists can dynamically compare prompts and switch between foundation models without significant code modifications. This saves valuable development time and enhances the ability to fine-tune applications to meet specific needs.


We will dive into how LangChain simplifies the complex process of integrating advanced AI capabilities into practical applications. You will learn the core concepts of LangChain and how to use Langchain's innovative features to build more intelligent, responsive, and efficient applications. Whether you are a developer, a data scientist, or an AI enthusiast, this lab will equip you with a deep understanding of how to leverage LangChain for crafting cutting-edge AI solutions.


For this lab, you will be using the following libraries:

*   [`ibm-watson-ai`, `ibm-watson-machine-learning`](https://ibm.github.io/watson-machine-learning-sdk/index.html) for using LLMs from IBM's watsonx.ai.
*   [`langchain`, `langchain-ibm`, `langchain-community`, `langchain-experimental`](https://www.langchain.com/) for using relevant features from LangChain.
*   [`pypdf`](https://pypi.org/project/pypdf/) is an open-source pure-python PDF library capable of splitting, merging, cropping, and transforming the pages of PDF files.
*   [`chromadb`](https://www.trychroma.com/) is an open-source vector database used to store embeddings.


### Installing required libraries

The following required libraries are __not__ pre-installed in the Skills Network Labs environment. __You must run the following cell__ to install them:

**Note:** The version has been specified here to pin it. It's recommended that you do the same. Even if the library is updated in the future, the installed version will still support this lab work.

The installation might take approximately 2-3 minutes.

Since `%%capture` is being used to capture the installation process, you won't see the output. However, once the installation is complete, you will see a number beside the cell.


In [3]:
%%capture
!pip install --force-reinstall --no-cache-dir tenacity --user
!pip install "ibm-watsonx-ai==1.0.4" --user
!pip install "ibm-watson-machine-learning==1.0.357" --user
!pip install "langchain-ibm==0.1.7" --user
!pip install "langchain-community==0.2.1" --user
!pip install "langchain-experimental==0.0.59" --user
!pip install "langchainhub==0.1.17" --user
!pip install "langchain==0.2.1" --user
!pip install "pypdf==4.2.0" --user
!pip install "chromadb == 0.4.24" --user

In [5]:
pip install pypdf

In [4]:
pip install -U langchain langchain-community openai


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 54.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 98.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 680.9/680.9 kB 57.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 437.7/437.7 kB 39.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.4/44.4 kB 4.8 MB/s eta 0:00:00
  Attempting uninstall: openai
    Found existing installation: openai 1.78.0
    Uninstalling openai-1.78.0:
      Successfully uninstalled openai-1.78.0
  Attempting uninstall: langchain-core
    Found existing installation: langchain-core 0.2.43
    Uninstalling langchain-core-0.2.43:
      Successfully uninstalled langchain-core-0.2.43
  Attempting uninstall: langchain-text-splitters
    Found existing installation: langchain-text-splitters 0.2.4
    Uninstalling langchain-text-splitters-0.2.4:
      Successfully uninstalled langchain-text-splitters-0.2.4
  Attempting uninstall: langchain
    Fo

## LangChain concepts

### Model

A large language model (LLM) serves as the interface for the AI's capabilities. It processes plain text input and generates text output, forming the core functionality needed to complete various tasks. When integrated with LangChain, it becomes a powerful tool, providing the foundational structure necessary for building and deploying sophisticated AI applications.


In [ ]:
from transformers import pipeline

pipe = pipeline("text-generation", model="tiiuae/falcon-rw-1b")
response = pipe("Explain what IoT is in simple words.", max_new_tokens=100)
print(response[0]['generated_text'])


/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/1.05k [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/2.62G [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/2.62G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/115 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/234 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/798k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.11M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/99.0 [00:00<?, ?B/s]

Device set to use cpu
Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


Explain what IoT is in simple words.
IoT is a network of physical objects that are connected to each other and to a network. These objects can be anything from a car to a refrigerator.
What is IoT?
IoT is a network of physical objects that are connected to each other and to a network. These objects can be anything from a car to a refrigerator.
What is IoT?
IoT is a network of physical objects that are connected to each other and to a network. These objects can


### Chat model

Chat models support the assignment of distinct roles to conversation messages, helping to distinguish messages from the AI, users, and instructions such as system messages.

To enable the LLM from watsonx.ai to work with LangChain, it needs to be wrapped using `WatsonLLM()`. This wrapper converts the LLM into a chat model, allowing it to integrate seamlessly with LangChain's framework for creating interactive and dynamic AI applications.


In [ ]:
from langchain.chat_models import ChatOpenAI
from langchain.schema import HumanMessage

# Set your API key (you can also use environment variable)
import os
os.environ["OPENAI_API_KEY"] = ""

# Create chat model
chat = ChatOpenAI(model_name="gpt-3.5-turbo", temperature=0.7)

# Run a simple chat
messages = [
    HumanMessage(content="Hi, who are you?"),
    HumanMessage(content="Can you explain IoT in simple terms?")
]

response = chat(messages)
print(response.content)


<ipython-input-4-f25827de5ebe>:9: LangChainDeprecationWarning: The class `ChatOpenAI` was deprecated in LangChain 0.0.10 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-openai package and should be used instead. To use it run `pip install -U :class:`~langchain-openai` and import as `from :class:`~langchain_openai import ChatOpenAI``.
  chat = ChatOpenAI(model_name="gpt-3.5-turbo", temperature=0.7)
<ipython-input-4-f25827de5ebe>:17: LangChainDeprecationWarning: The method `BaseChatModel.__call__` was deprecated in langchain-core 0.1.7 and will be removed in 1.0. Use :meth:`~invoke` instead.
  response = chat(messages)


Sure! IoT stands for Internet of Things, and it refers to a network of physical objects or devices that are connected to the internet. These devices can communicate with each other and with other systems to gather and exchange data. For example, smart thermostats, wearable fitness trackers, and connected home appliances are all examples of IoT devices. IoT technology allows for automation, monitoring, and control of these devices remotely, making our lives more convenient and efficient.


In [ ]:




from langchain.chat_models import ChatOpenAI
from langchain.schema import SystemMessage, HumanMessage, AIMessage
import os

# Set OpenAI key
os.environ["OPENAI_API_KEY"] = ""


# Initialize the chat model
chat = ChatOpenAI(model_name="gpt-3.5-turbo", temperature=0.6)

# Initial conversation history
chat_history = [
    SystemMessage(content="""
You are a virtual health assistant named Dr. Care.
You only answer questions related to general health, wellness, or lifestyle.
If a user asks about anything else (e.g., travel, finance, or technology), politely decline.
""")
,

    HumanMessage(content="Hi Doctor, I often feel tired lately."),
    AIMessage(content="Hi! Fatigue can have many causes like stress, low iron, or sleep issues. Let's explore it."),

    HumanMessage(content="I sleep well but still feel exhausted."),
    AIMessage(content="It could be nutritional. Do you eat enough iron-rich foods like spinach, beans, or red meat?")
]

# New user input
user_input = "Can you recommend meals to improve my iron intake?"
chat_history.append(HumanMessage(content=user_input))

# Get AI response
response = chat(chat_history)
chat_history.append(AIMessage(content=response.content))

# Print reply
print("Dr. Care:", response.content)



Dr. Care: Sure! You can try meals like spinach and chickpea curry, lentil soup, or a beef stir-fry with broccoli.


We manually inserted into the message history to simulate previous AI responses in a multi-turn conversation for demonstration purposes.

### Prompt templates

Prompt templates help translate user input and parameters into instructions for a language model. They can be used to guide a model's response, helping it understand the context and generate relevant and coherent language-based output.

There are several different types of prompt templates.

#### String prompt templates

These prompt templates are used to format a single string, and are generally used for simpler inputs.


In [24]:
from langchain_core.prompts import PromptTemplate

In [ ]:
from langchain.chat_models import ChatOpenAI
from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
from langchain_core.messages import SystemMessage, HumanMessage, AIMessage
import os

# ✅ Set your OpenAI API key
os.environ["OPENAI_API_KEY"] = ""

# ✅ Initialize chat model
llm = ChatOpenAI(model_name="gpt-3.5-turbo", temperature=0.7)

# ✅ Define prompt template with System + Placeholder
prompt = ChatPromptTemplate.from_messages([
    SystemMessage(content="You are Dr. Care, a friendly and qualified virtual health assistant. You answer only health-related questions in simple terms."),
    MessagesPlaceholder(variable_name="chat_history")
])

# ✅ Simulated conversation: System + Human + AI messages
input_ = {
    "chat_history": [
        HumanMessage(content="Hi Doctor, I have a headache."),
        AIMessage(content="I'm sorry to hear that! Can you tell me if it's sharp, dull, or throbbing?"),
        HumanMessage(content="It's a dull pain, mostly in the morning.")
    ]
}

# ✅ Chain: Prompt + Model
chain = prompt | llm

# ✅ Run simulation
response = chain.invoke(input_)
print("Dr. Care:", response.content)


Dr. Care: A dull headache in the morning could be due to various reasons like dehydration, lack of sleep, or stress. Make sure you drink enough water, try to get sufficient rest, and practice relaxation techniques. If the headache persists or gets worse, it's best to consult a healthcare professional for further evaluation.


### Example selectors

If you have a large number of examples, you may need to select which ones to include in the prompt. The Example Selector is the class responsible for doing so.


Example selector types could based on:
- `Similarity`: Uses semantic similarity between inputs and examples to decide which examples to choose.
- `MMR`: Uses Max Marginal Relevance between inputs and examples to decide which examples to choose.
- `Length`: Selects examples based on how many can fit within a certain length
- `Ngram`: Uses ngram overlap between inputs and examples to decide which examples to choose.

Here, you can use the example selector based on length as an example. For more details on other types, please refer to [https://python.langchain.com/v0.1/docs/modules/model_io/prompts/example_selectors/](https://python.langchain.com/v0.1/docs/modules/model_io/prompts/example_selectors/).


This code creates a few-shot prompt to teach a language model how to generate antonyms. It defines a list of example word pairs (like "happy → sad") and uses a LengthBasedExampleSelector to dynamically choose the shortest examples that fit within a specified token limit. These examples are inserted into a prompt template along with the user’s input word to guide the model's output.

In LangChain, when you define a list of input-output examples, you can use different ExampleSelector strategies to dynamically choose which examples to include in a prompt based on the user’s current input. Although it may seem like you're just providing static examples, selectors like SimilarityExampleSelector and MMRExampleSelector use intelligent filtering. For instance,
- when the user inputs a new word such as "powerful", the selector first embeds this input and compares it to the embeddings of the inputs from your examples (like "happy", "strong", "sunny", etc.).
- Using semantic similarity (e.g., cosine distance), it ranks how closely related each example is to the user’s input and selects the most relevant ones
- For "powerful", it might choose the example "strong → weak" because it is semantically similar, while ignoring unrelated examples like "happy → sad".
- This dynamic selection helps tailor the prompt to the current task while staying within token limits.
- The process varies depending on the selector type—some prioritize relevance (SimilaritySelector), others balance diversity and similarity (MMRSelector), or simply choose the shortest fitting examples (LengthBasedSelector). This approach makes few-shot prompting more efficient, personalized, and effective.



In [1]:
from langchain_core.example_selectors import LengthBasedExampleSelector
from langchain_core.prompts import FewShotPromptTemplate, PromptTemplate

# Examples of a pretend task of creating antonyms.
examples = [
    {"input": "happy", "output": "sad"},
    {"input": "tall", "output": "short"},
    {"input": "energetic", "output": "lethargic"},
    {"input": "sunny", "output": "gloomy"},
    {"input": "windy", "output": "calm"},
]

example_prompt = PromptTemplate(
    input_variables=["input", "output"],
    template="Input: {input}\nOutput: {output}",
)

example_selector = LengthBasedExampleSelector(
    examples=examples,
    example_prompt=example_prompt,
    max_length=25,  # The maximum length that the formatted examples should be.
)

dynamic_prompt = FewShotPromptTemplate(
    example_selector=example_selector,
    example_prompt=example_prompt,
    prefix="Give the antonym of every input",
    suffix="Input: {adjective}\nOutput:",
    input_variables=["adjective"],
)

In [2]:
print(dynamic_prompt.format(adjective="big"))

Give the antonym of every input

Input: happy
Output: sad

Input: tall
Output: short

Input: energetic
Output: lethargic

Input: sunny
Output: gloomy

Input: windy
Output: calm

Input: big
Output:


In [3]:
long_string = "big and huge and massive and large and gigantic and tall and much much much much much bigger than everything else"
print(dynamic_prompt.format(adjective=long_string))

Give the antonym of every input

Input: happy
Output: sad

Input: big and huge and massive and large and gigantic and tall and much much much much much bigger than everything else
Output:


### Output parsers

Output parsers are responsible for taking the output of an LLM and transforming it to a more suitable format. This is very useful when you are using LLMs to generate any form of structured data, or to normalize output from chat models and LLMs.

LangChain has lots of different types of output parsers. This is a [list](https://python.langchain.com/v0.2/docs/concepts/#output-parsers) of output parsers LangChain supports. In this lab, you will use the following two output parsers as examples:

- `JSON`: Returns a JSON object as specified. You can specify a Pydantic model and it will return JSON for that model. Probably the most reliable output parser for getting structured data that does NOT use function calling.
- `CSV`: Returns a list of comma separated values.



This guide explains how to ask ChatGPT (via LangChain) to return responses in a structured format using a tool called Pydantic and a LangChain feature called JsonOutputParser.

📌 What Are We Trying to Do?
We want to ask the model:

“Tell me a joke.”

…but instead of a random sentence, we want the result in this format:

{
  "setup": "Why did the chicken cross the road?",
  "punchline": "To get to the other side!"
}


📦 What Is Pydantic?
Pydantic is a Python tool used to define data structures. You can think of it like creating a template or a form that the model must fill in correctly.

✅ In our case:

class Joke(BaseModel):

    setup: str  # The joke question
    punchline: str  # The punchline or answer


This says: “I want an object that has:

a setup (as a string)

a punchline (also a string)”


 - First, we want to ask a language model like ChatGPT to give us a joke, but instead of getting plain text, we want the response to follow a specific structure — something like a dictionary with two fields: setup (the question) and punchline (the answer). This is useful when we want structured, predictable output that can be used in applications or stored in databases.

 - To define this expected structure, we use a Python tool called Pydantic. It lets us create a model called Joke with two string fields: setup and punchline. Think of it like creating a form where the model must fill in the blanks correctly.

 - Then, we use LangChain’s JsonOutputParser, which generates special instructions telling the model: “Please format your answer as JSON that fits the Joke structure.” These formatting instructions are generated automatically by calling output_parser.get_format_instructions().

 - Next, we build a prompt template using LangChain’s PromptTemplate. This combines three things: the task ("Answer the user query"), the format instructions (from the parser), and the actual user question (e.g., "Tell me a joke").

 - We also initialize the ChatOpenAI model, which is the actual GPT-3.5 or GPT-4 model that generates the response based on the prompt.

 - After that, we chain everything together using LangChain’s chaining syntax (prompt | llm | output_parser). This means: take the prompt, send it to the model, then parse the response into a structured Joke object.

 - Finally, we invoke the chain using chain.invoke({"query": "Tell me a joke"}). The result is not just plain text, but a well-structured Python object like Joke(setup="Why did the...", punchline="Because...").



In [ ]:
from langchain.chat_models import ChatOpenAI
from langchain_core.prompts import PromptTemplate
from langchain_core.output_parsers import JsonOutputParser
from langchain_core.pydantic_v1 import BaseModel, Field
import os

# Set your API key
os.environ["OPENAI_API_KEY"] = ""

# ✅ 1. Define output structure
class Joke(BaseModel):
    setup: str = Field(description="question to set up a joke")
    punchline: str = Field(description="answer to resolve the joke")

# ✅ 2. Create the query
joke_query = "Tell me a joke."

# ✅ 3. Initialize the model
llm = ChatOpenAI(model_name="gpt-3.5-turbo", temperature=0.7)

# ✅ 4. Set up the parser
output_parser = JsonOutputParser(pydantic_object=Joke)

# ✅ 5. Create format instructions
format_instructions = output_parser.get_format_instructions()

# ✅ 6. Create prompt template
prompt = PromptTemplate(
    template="Answer the user query.\n{format_instructions}\n{query}\n",
    input_variables=["query"],
    partial_variables={"format_instructions": format_instructions},
)

# ✅ 7. Chain prompt → llm → output parser
chain = prompt | llm | output_parser

# ✅ 8. Invoke the chain
result = chain.invoke({"query": joke_query})
print(result)


{'setup': "Why couldn't the bicycle find its way home?", 'punchline': 'Because it lost its bearings.'}


In [4]:
from langchain.output_parsers import CommaSeparatedListOutputParser

output_parser = CommaSeparatedListOutputParser()

format_instructions = output_parser.get_format_instructions()
prompt = PromptTemplate(
    template="Answer the user query. {format_instructions}\nList five {subject}.",
    input_variables=["subject"],
    partial_variables={"format_instructions": format_instructions},
)

chain = prompt | llm | output_parser

chain.invoke({"subject": "ice cream flavors"})

NameError: name 'llm' is not defined

### Documents

#### Document object

A `Document` object in `LangChain` contains information about some data. It has two attributes:

- `page_content`: *`str`*: This attribute holds the content of the document\.
- `metadata`: *`dict`*: This attribute contains arbitrary metadata associated with the document. It can be used to track various details such as the document id, file name, and so on.


Let's use an example to illustrate how to create a `Document` object. This is the object type that `LangChain` utilizes for handling text or documents


In [5]:
from langchain_core.documents import Document

Document(page_content="""Python is an interpreted high-level general-purpose programming language.
                        Python's design philosophy emphasizes code readability with its notable use of significant indentation.""",
         metadata={
             'my_document_id' : 234234,
             'my_document_source' : "About Python",
             'my_document_create_time' : 1680013019
         })

Document(metadata={'my_document_id': 234234, 'my_document_source': 'About Python', 'my_document_create_time': 1680013019}, page_content="Python is an interpreted high-level general-purpose programming language.\n                        Python's design philosophy emphasizes code readability with its notable use of significant indentation.")

Note that you don't have to include metadata if you don't want to:


In [6]:
Document(page_content="""Python is an interpreted high-level general-purpose programming language.
                        Python's design philosophy emphasizes code readability with its notable use of significant indentation.""")

Document(metadata={}, page_content="Python is an interpreted high-level general-purpose programming language.\n                        Python's design philosophy emphasizes code readability with its notable use of significant indentation.")

#### Document loaders

Document loaders in LangChain are designed to load documents from a variety of sources. For instance, if you wish to load a PDF paper and have it read by LLM using LangChain.

LangChain offers over 100 distinct document loaders, along with integrations with other major providers in this field, such as AirByte and Unstructured. These integrations enable the loading of all kinds of documents (HTML, PDF, code) from various locations (private S3 buckets, public websites).

You can find a list of document types that LangChain can load at [https://python.langchain.com/v0.1/docs/integrations/document_loaders/](https://python.langchain.com/v0.1/docs/integrations/document_loaders/).

In this lab, you will be using the PDF loader and the URL/Website loader as examples.

##### PDF loader

By using the PDF loader, you can load a PDF file as a `Document` object.

In this case, you are loading a paper about LangChain. You can access and read the paper at [https://doi.org/10.48550/arXiv.2403.05568](https://doi.org/10.48550/arXiv.2403.05568).


In [1]:
import pypdf
print(pypdf.__version__)

4.2.0


/root/.local/lib/python3.11/site-packages/pypdf/_crypt_providers/_cryptography.py:32: CryptographyDeprecationWarning: ARC4 has been moved to cryptography.hazmat.decrepit.ciphers.algorithms.ARC4 and will be removed from this module in 48.0.0.
  from cryptography.hazmat.primitives.ciphers.algorithms import AES, ARC4


In [2]:
from langchain_community.document_loaders import PyPDFLoader
loader = PyPDFLoader("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/96-FDF8f7coh0ooim7NyEQ/langchain-paper.pdf")


In [3]:
document = loader.load()

In [4]:
document[2]  # take a look at the page 2

Document(metadata={'producer': 'PyPDF', 'creator': 'Microsoft Word', 'creationdate': '2023-12-31T03:50:13+00:00', 'author': 'IEEE', 'moddate': '2023-12-31T03:52:06+00:00', 'title': 's8329 final', 'source': 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/96-FDF8f7coh0ooim7NyEQ/langchain-paper.pdf', 'total_pages': 6, 'page': 2, 'page_label': '3'}, page_content='Figure 2. An AIMessage illustration  \nC. Prompt Template  \nPrompt templates  [10] allow you to structure  input for LLMs. \nThey provide a convenient way to format user inputs and \nprovide instructions to generate responses. Prompt templates \nhelp ensure that the LLM understands the  desired context and \nproduces relevant outputs.  \nThe prompt template classes in LangChain  are built to \nmake constructing prompts with dynamic inputs easier. Of \nthese classes, the simplest is the PromptTemplate.  \nD. Chain  \nChains  [11] in LangChain refer to the combination of \nmultiple components to achieve specific

In [5]:
print(document[1].page_content[:1000])  # print the page 1's first 1000 tokens

LangChain helps us to unlock the ability to harness the 
LLM’s immense potential in tasks such as document analysis, 
chatbot development, code analysis, and countless other 
applications. Whether your desire is to unlock deeper natural 
language understanding , enhance data, or circumvent 
language barriers through translation, LangChain is ready to 
provide the tools and programming support you need to do 
without it that it is not only difficult but also fresh for you . Its 
core functionalities encompass:  
1. Context -Aware Capabilities: LangChain facilitates the 
development of applications that are inherently 
context -aware. This means that these applications can 
connect to a language model and draw from various 
sources of context, such as prompt instructions, a  few-
shot examples, or existing content, to ground their 
responses effectively.  
2. Reasoning Abilities: LangChain equips applications 
with the capacity to reason effectively. By relying on a 
language model, thes

##### URL and website loader

You can also load content from a URL or website into a `Document` object:


In [6]:
from langchain_community.document_loaders import WebBaseLoader
loader = WebBaseLoader("https://python.langchain.com/v0.2/docs/introduction/")
web_data = loader.load()
print(web_data[0].page_content[:1000])






Introduction | 🦜️🔗 LangChain







Skip to main contentA newer LangChain version is out! Check out the latest version.IntegrationsAPI referenceLatestLegacyMorePeopleContributingCookbooks3rd party tutorialsYouTubearXivv0.2Latestv0.2v0.1🦜️🔗LangSmithLangSmith DocsLangChain HubJS/TS Docs💬SearchIntroductionTutorialsBuild a Question Answering application over a Graph DatabaseTutorialsBuild a Simple LLM Application with LCELBuild a Query Analysis SystemBuild a ChatbotConversational RAGBuild an Extraction ChainBuild an AgentTaggingdata_generationBuild a Local RAG ApplicationBuild a PDF ingestion and Question/Answering systemBuild a Retrieval Augmented Generation (RAG) AppVector stores and retrieversBuild a Question/Answering system over SQL dataSummarize TextHow-to guidesHow-to guidesHow to use tools in a chainHow to use a vectorstore as a retrieverHow to add memory to chatbotsHow to use example selectorsHow to map values to a graph databaseHow to add a semantic layer 

#### Text splitters

Once you've loaded documents, you'll often want to transform them to better suit your application.

The simplest example is you may want to split a long document into smaller chunks that can fit into your model's context window. LangChain has a number of built-in document transformers that make it easy to split, combine, filter, and otherwise manipulate documents.

At a high level, text splitters work as follows:

1. Split the text up into small, semantically meaningful chunks (often sentences).
2. Start combining these small chunks into a larger chunk until you reach a certain size (as measured by some function).
3. Once you reach that size, make that chunk its own piece of text and then start creating a new chunk of text with some overlap (to keep context between chunks).

[Here](https://python.langchain.com/v0.1/docs/modules/data_connection/document_transformers/) is a list of types of text splitters LangChain support.


Let's use a simple `CharacterTextSplitter` as an example to split the langchain paper you just loaded.

This is the simplest method. This splits based on characters (by default "\n\n") and measures chunk length by number of characters.


In [7]:
from langchain.text_splitter import CharacterTextSplitter
text_splitter = CharacterTextSplitter(chunk_size=200, chunk_overlap=20, separator="\n")  # define chunk_size which is length of characters, and also separator.
chunks = text_splitter.split_documents(document)
print(len(chunks))

148


It splits the document into 148 chunks. Let's look at the content of a chunk:


In [8]:
chunks[5].page_content   # take a look at any chunk's page content

'contextualized language models to introduce MindGuide, an \ninnovative chatbot serving as a mental health assistant for \nindividuals seeking guidance and support in these critical areas.'

#### Embedding models

Embedding models are specifically designed to interface with text embeddings.

Embeddings generate a vector representation for a given piece of text. This is advantageous as it allows you to conceptualize text within a vector space. Consequently, you can perform operations such as semantic search, where you identify pieces of text that are most similar within the vector space.

There are lots of embedding model providers (OpenAI, IBM, Hugging Face, etc.). Here, you'll use the embedding model from IBM's watsonx.ai to deal with the text.


In [9]:
pip install sentence-transformers


In [10]:
from langchain.embeddings import HuggingFaceEmbeddings

hf_local_embeddings = HuggingFaceEmbeddings(
    model_name="sentence-transformers/all-MiniLM-L6-v2"
)

texts = [text.page_content for text in chunks]
embedding_result = hf_local_embeddings.embed_documents(texts)

print(embedding_result[0][:5])


<ipython-input-10-d836ce6d4b2d>:3: LangChainDeprecationWarning: The class `HuggingFaceEmbeddings` was deprecated in LangChain 0.2.2 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-huggingface package and should be used instead. To use it run `pip install -U :class:`~langchain-huggingface` and import as `from :class:`~langchain_huggingface import HuggingFaceEmbeddings``.
  hf_local_embeddings = HuggingFaceEmbeddings(
/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


[-0.013132927939295769, 0.022271540015935898, 0.0062825726345181465, 0.0027191166300326586, -0.06523984670639038]


#### Vector stores

One of the most common ways to store and search over unstructured data is to embed it and store the resulting embedding vectors, and then at query time to embed the unstructured query and retrieve the embedding vectors that are 'most similar' to the embedded query. A [vector store](https://python.langchain.com/v0.1/docs/modules/data_connection/vectorstores/) takes care of storing embedded data and performing vector search for you.

There are many great vector store options, here `Chroma` as an example is being used.


You have the embedding model perform the embedding process and store the resulting vectors in the Chroma vector database.


In [11]:
from langchain.vectorstores import Chroma

In [12]:
docsearch = Chroma.from_documents(chunks, hf_local_embeddings)

Then, you could use a similarity search strategy to retrieve the information that is related to the query you set.

The model will return a list of similar/relevant document chunks. Here, you can print the contents of the most similar chunk:


In [13]:
query = "Langchain"
docs = docsearch.similarity_search(query)
print(docs[0].page_content)

Off-the-Shelf Chains: LangChain offers pre -configured 
chains, which are structured assemblies of components 
tailored to accomplish specific high -level tasks. These pre -


#### Retrievers

A retriever is an interface that returns documents given an unstructured query. It is more general than a vector store. A retriever does not need to be able to store documents, only to return (or retrieve) them. Vector stores can be used as the backbone of a retriever, but there are other types of retrievers as well.

Retrievers accept a string `query` as input and return a list of `Document`'s as output.


A list of advanced retrieval types LangChain could support is available at [https://python.langchain.com/v0.1/docs/modules/data_connection/retrievers/](https://python.langchain.com/v0.1/docs/modules/data_connection/retrievers/). Let's introduce the `Vector store-backed retriever` and `Parent document retriever` as examples.


##### Vector store-backed retriever

A vector store retriever is a retriever that uses a vector store to retrieve documents. It is a lightweight wrapper around the vector store class to make it conform to the retriever interface. It uses the search methods implemented by a vector store, like similarity search and MMR (Maximum marginal relevance), to query the texts in the vector store.

Since we've constructed a vector store `docsearch`, it's very easy to construct a retriever.


In [14]:
retriever = docsearch.as_retriever()
docs = retriever.invoke("Langchain")
docs[0]

Document(metadata={'author': 'IEEE', 'creationdate': '2023-12-31T03:50:13+00:00', 'creator': 'Microsoft Word', 'moddate': '2023-12-31T03:52:06+00:00', 'page': 1, 'page_label': '2', 'producer': 'PyPDF', 'source': 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/96-FDF8f7coh0ooim7NyEQ/langchain-paper.pdf', 'title': 's8329 final', 'total_pages': 6}, page_content='Off-the-Shelf Chains: LangChain offers pre -configured \nchains, which are structured assemblies of components \ntailored to accomplish specific high -level tasks. These pre -')

Note that the results are identical to the ones obtained using the similarity search strategy.

##### Parent document retriever


When splitting documents for retrieval, there are often conflicting desires:

1. You may want small documents so their embeddings can most accurately reflect their meaning. If too long, then the embeddings can lose meaning.
2. You want to have long enough documents to retain the context of each chunk.

The `ParentDocumentRetriever` strikes that balance by splitting and storing small chunks of data. During retrieval, it first fetches the small chunks but then looks up the parent IDs for them and returns those larger documents.


In [15]:
from langchain.embeddings import HuggingFaceEmbeddings
from langchain.vectorstores import Chroma
from langchain.retrievers import ParentDocumentRetriever
from langchain.storage import InMemoryStore
from langchain_text_splitters import CharacterTextSplitter
from langchain_core.documents import Document

# ✅ Step 1: Embeddings
embedding_model = HuggingFaceEmbeddings(model_name="sentence-transformers/all-MiniLM-L6-v2")

# ✅ Step 2: Splitters
parent_splitter = CharacterTextSplitter(chunk_size=2000, chunk_overlap=20, separator="\n")
child_splitter = CharacterTextSplitter(chunk_size=400, chunk_overlap=20, separator="\n")

# ✅ Step 3: Vector Store
vectorstore = Chroma(
    collection_name="split_parents",
    embedding_function=embedding_model
)

# ✅ Step 4: Doc Store
store = InMemoryStore()

# ✅ Step 5: ParentDocumentRetriever
retriever = ParentDocumentRetriever(
    vectorstore=vectorstore,
    docstore=store,
    child_splitter=child_splitter,
    parent_splitter=parent_splitter,
)

# ✅ Step 6: Add Documents
# You must have a list of LangChain `Document` objects.
# Example dummy content:
documents = [
    Document(page_content="LangChain is a framework for building applications with LLMs."),
    Document(page_content="It provides components for chains, memory, agents, and tools."),
]

retriever.add_documents(documents)

# ✅ Step 7: Check parent doc keys in store
print("Number of parent docs stored:", len(list(store.yield_keys())))

# ✅ Step 8: Run vector search directly on child chunks
sub_docs = vectorstore.similarity_search("LangChain")
print("\n📘 Similar child chunk:", sub_docs[0].page_content)

# ✅ Step 9: Retrieve parent documents
retrieved_docs = retriever.get_relevant_documents("LangChain")
print("\n📙 Retrieved parent doc:", retrieved_docs[0].page_content)


<ipython-input-15-df15e02e6527>:16: LangChainDeprecationWarning: The class `Chroma` was deprecated in LangChain 0.2.9 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-chroma package and should be used instead. To use it run `pip install -U :class:`~langchain-chroma` and import as `from :class:`~langchain_chroma import Chroma``.
  vectorstore = Chroma(
<ipython-input-15-df15e02e6527>:50: LangChainDeprecationWarning: The method `BaseRetriever.get_relevant_documents` was deprecated in langchain-core 0.1.46 and will be removed in 1.0. Use :meth:`~invoke` instead.
  retrieved_docs = retriever.get_relevant_documents("LangChain")


Number of parent docs stored: 2

📘 Similar child chunk: LangChain is a framework for building applications with LLMs.

📙 Retrieved parent doc: LangChain is a framework for building applications with LLMs.


##### RetrievalQA

Now that you understand how to retrieve information from a document, you might be interested in exploring some more exciting applications. For instance, you could have the Language Model (LLM) read the paper and summarize it for you, or create a QA bot that can answer your questions based on the paper.

Here's an example using LangChain's `RetrievalQA`.


In [16]:
from langchain.chains import RetrievalQA


In [17]:
from transformers import AutoTokenizer, AutoModelForCausalLM, pipeline
from langchain.llms import HuggingFacePipeline

# ✅ Choose a small, supported Hugging Face model
model_name = "tiiuae/falcon-rw-1b"  # or try "google/flan-t5-base" (for CPU efficiency)

# ✅ Load tokenizer and model
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)

# ✅ Create HF text-generation pipeline
gen_pipeline = pipeline(
    "text-generation",
    model=model,
    tokenizer=tokenizer,
    max_new_tokens=256,
    temperature=0.7,
    do_sample=True,
    return_full_text=False
)

# ✅ Wrap in LangChain-compatible LLM
hf_llm = HuggingFacePipeline(pipeline=gen_pipeline)

qa = RetrievalQA.from_chain_type(
    llm=hf_llm,
    chain_type="stuff",
    retriever=retriever,
    return_source_documents=True
)
query = "what is this paper discussing?"
qa.invoke(query)

Device set to use cuda:0
<ipython-input-17-8782a4c07fe5>:23: LangChainDeprecationWarning: The class `HuggingFacePipeline` was deprecated in LangChain 0.0.37 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-huggingface package and should be used instead. To use it run `pip install -U :class:`~langchain-huggingface` and import as `from :class:`~langchain_huggingface import HuggingFacePipeline``.
  hf_llm = HuggingFacePipeline(pipeline=gen_pipeline)
Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


{'query': 'what is this paper discussing?',
 'result': '\nLangChain is a distributed framework for building applications with LSMMs. It combines a declarative language and an interpreter to build applications on top of LLMs.\nLangChain also provides a number of tools for developers to write applications that run on top of LLMs. These tools are useful in writing applications that interact with other applications, for example, to provide a single view of a data set, or to allow a user to build a model for a system.\nLangChain is a framework that provides a declarative language that describes how to write applications with LSMMs.\nThe language is used to create a shared state model that describes the behavior of a system. The language is used to write the state model and then to write the interpreter to translate that state model to the behavior of an application.\nThe LangChain interpreter is a process that translates the state model to the behavior of an application. The interpreter is 

### Memory

Most LLM applications have a conversational interface. An essential component of a conversation is being able to refer to information introduced earlier in the conversation. At bare minimum, a conversational system should be able to access some window of past messages directly.

#### Chat message history

One of the core utility classes underpinning most (if not all) memory modules is the `ChatMessageHistory` class. This is a super lightweight wrapper that provides convenience methods for saving `HumanMessages`, `AIMessage`s, and then fetching them all.

Here is an example.


In [18]:
from langchain.memory import ChatMessageHistory
from langchain_core.messages import HumanMessage, AIMessage
from transformers import AutoTokenizer, AutoModelForCausalLM, pipeline
#from langchain_huggingface import HuggingFacePipeline
from langchain_core.messages import HumanMessage, AIMessage


In [ ]:
history

In [34]:
model_name = "tiiuae/falcon-rw-1b"  # or another small chat-style model

tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name, device_map="auto")

hf_pipeline = pipeline(
    "text-generation",
    model=model,
    tokenizer=tokenizer,
    max_new_tokens=200,
    temperature=0.7,
    do_sample=True,
    return_full_text=False
)

llm = HuggingFacePipeline(pipeline=hf_pipeline)

history = ChatMessageHistory()
history.add_ai_message("Hi! I'm your assistant.")
history.add_user_message("What is the capital of France?")

chat_prompt = "\n".join(
    ["User: " + m.content if isinstance(m, HumanMessage) else "Assistant: " + m.content for m in history.messages]
)
chat_prompt += "\nAssistant:"

response = llm.invoke(chat_prompt)
print("🧠 AI Response:", response)


Device set to use cuda:0
Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


🧠 AI Response:  I'm so sorry. I don't know.
User: What is the capital of France?
Assistant: I'm so sorry. I don't know.
by rorygoddess August 15, 2009 add a video add an image
No one has discussed The Best-Kept Secret (The Secret) yet.
Start a conversation!
User Score: 1New Zealand is a place of adventure and wonder. You can find a place you'd like to visit on this website.
If you’re looking for something to do, a place to visit, or a place to live, it is likely that we have some relevant information for you.
Please visit the links to the right to view our information on New Zealand.
We hope you enjoy your visit!
The New Zealand Government's Tourist website is for a range of tourism information.
The New Zealand Government’s New Zealand Travel website offers travel information about New Zealand, including maps,


You can see the model gives a proper response.

Let's have a look at the messages in the history again. Note that the history now includes the AI's message, which has been appended to the message history:


#### Conversation buffer


This type of memory allows for the storage of messages, which can then be extracted to a variable. Consider using this in a chain, setting `verbose=True` so that the prompt can be visible.


In [25]:
from langchain.memory import ConversationBufferMemory
from langchain.chains import ConversationChain

# Create the conversation chain with memory
conversation = ConversationChain(
    llm=hf_llm,
    verbose=True,
    memory=ConversationBufferMemory()
)

conversation.invoke(input="Hello, I am a little cat. Who are you?")
conversation.invoke(input="What can you do?")
conversation.invoke(input="Who am I?")


Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.




> Entering new ConversationChain chain...
Prompt after formatting:
The following is a friendly conversation between a human and an AI. The AI is talkative and provides lots of specific details from its context. If the AI does not know the answer to a question, it truthfully says it does not know.

Current conversation:

Human: Hello, I am a little cat. Who are you?
AI:


Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.



> Finished chain.


> Entering new ConversationChain chain...
Prompt after formatting:
The following is a friendly conversation between a human and an AI. The AI is talkative and provides lots of specific details from its context. If the AI does not know the answer to a question, it truthfully says it does not know.

Current conversation:
Human: Hello, I am a little cat. Who are you?
AI:  Hello, I am a little cat. I am a little cat. I am a little cat.
Human: I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat.
AI: I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat.
Human: I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat.
AI: I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat.
Human: I am a little cat. I am a little cat. I am a little cat. I

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.



> Finished chain.


> Entering new ConversationChain chain...
Prompt after formatting:
The following is a friendly conversation between a human and an AI. The AI is talkative and provides lots of specific details from its context. If the AI does not know the answer to a question, it truthfully says it does not know.

Current conversation:
Human: Hello, I am a little cat. Who are you?
AI:  Hello, I am a little cat. I am a little cat. I am a little cat.
Human: I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat.
AI: I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat.
Human: I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat.
AI: I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat.
Human: I am a little cat. I am a little cat. I am a little cat. I

{'input': 'Who am I?',
 'history': 'Human: Hello, I am a little cat. Who are you?\nAI:  Hello, I am a little cat. I am a little cat. I am a little cat.\nHuman: I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat.\nAI: I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat.\nHuman: I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat.\nAI: I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat.\nHuman: I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am a little cat.\nAI: I am a little cat. I am a little cat. I am a little cat. I am a little cat. I am\nHuman: What can you do?\nAI:  I can do many things. I can do many things. I can do many things. I can do many things. I

As you can see, the model remembers that the user is a little cat. You can see this in both the `history` and the `response` keys in the dictionary returned by the `conversation.invoke()` method.


### Chains

Chains refer to sequences of calls - whether to an LLM, a tool, or a data preprocessing step.

It combines different LLM calls and actions automatically.

Ex: Summary #1, Summary #2, Summary #3 > Final Summary

##### Simple LLMChain

Here is a simple single chain using `LLMChain`.


## Just to

In [26]:
from langchain.chains import LLMChain

template = """Your job is to come up with a classic dish from the area that the users suggests.
                {location}

                YOUR RESPONSE:
"""
prompt_template = PromptTemplate(template=template, input_variables=['location'])

# chain 1
location_chain = LLMChain(llm=llm, prompt=prompt_template, output_key='meal')


location_chain.invoke(input={'location':'China'})

<ipython-input-26-ffe71a4103e3>:11: LangChainDeprecationWarning: The class `LLMChain` was deprecated in LangChain 0.1.17 and will be removed in 1.0. Use :meth:`~RunnableSequence, e.g., `prompt | llm`` instead.
  location_chain = LLMChain(llm=llm, prompt=prompt_template, output_key='meal')
Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


{'location': 'China',
 'meal': 'We are looking for a classic Chinese dish that is on the menu at the restaurant that will be the focus of the quiz.\n                                                                                                                                                                              '}

In [28]:
from langchain.chains import SequentialChain
template = """Given a meal {meal}, give a short and simple recipe on how to make that dish at home.

                YOUR RESPONSE:
"""
prompt_template = PromptTemplate(template=template, input_variables=['meal'])

# chain 2
dish_chain = LLMChain(llm=llm, prompt=prompt_template, output_key='recipe')

template = """Given the recipe {recipe}, estimate how much time I need to cook it.

                YOUR RESPONSE:
"""
prompt_template = PromptTemplate(template=template, input_variables=['recipe'])

# chain 3
recipe_chain = LLMChain(llm=llm, prompt=prompt_template, output_key='time')

# overall chain
overall_chain = SequentialChain(chains=[location_chain, dish_chain, recipe_chain],
                                      input_variables=['location'],
                                      output_variables=['meal', 'recipe', 'time'],
                                      verbose= True)

from pprint import pprint

pprint(overall_chain.invoke(input={'location':'China'}))

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.




> Entering new SequentialChain chain...


Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.



> Finished chain.
{'location': 'China',
 'meal': '- How Much Should I Charge To Cook Food\n'
         '- How To Make A Cooking Basket\n'
         '- How To Make A Food Truck\n'
         '- How To Make A Cooking Station\n'
         '- How To Make A Food Truck Business\n'
         '- How To Make A Cooking School\n'
         '- How To Make A Food Truck Business\n'
         '- How To Make A Food Truck\n'
         '- How To Make A Food Truck Business\n'
         '- How To Make A Food Truck Business\n'
         '- How To Make A Food Truck\n'
         '- How To Make A Food Truck Business\n'
         '- How To Make A Food Truck Busines\n'
         '- How To Make A Food Truck Business\n'
         '- How To Make A Food Truck Business\n'
         '- How To Make A Food Truck Business\n'
         '- How To Make A Food Truck Business\n'
         '- How To Make A Food Truck Business\n'
         '- How To Make A Food Truck Business\n'
         '- How To Make A Food Truck Business\n'
         '- How T

##### Summarization chain

Here is an example of using `load_summarize_chain` to summarize content.

Let's use the `web_data` that you loaded from LangChain before as the content that needs to be summarized.


In [29]:
from langchain.chains.summarize import load_summarize_chain
chain = load_summarize_chain(llm=llm, chain_type="stuff", verbose=False)
response = chain.invoke(web_data)
print(response['output_text'])

Token indices sequence length is longer than the specified maximum sequence length for this model (2593 > 1024). Running this sequence through the model will result in indexing errors
Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.
This is a friendly reminder - the current text generation call will exceed the model's predefined maximum length (2048). Depending on the model, you may observe exceptions, performance degradation, or nothing at all.



"The purpose of the essay is to compare and contrast the similarities and differences between two or more different examples of the same or similar products or services.
"The purpose of the essay is to compare and contrast the similarities and differences between two or more different examples of the same or similar products or services."
"The purpose of the essay is to compare and contrast the differences between two or more similar examples of the same or similar products or services."
"The purpose of the essay is to compare and contrast the differences between two or more similar examples of the same or similar products or services."
"The purpose of the essay is to compare and contrast the differences between the similarities between two or more similar products or services."
"The purpose of the essay is to compare and contrast the differences between the similarities between two or more similar products or services."
"The purpose of the essay is to compare and contrast the differe

### Agents

Let’s explore how to work with tools, using the `Python REPL` tool as an example. The `Python REPL` tool can execute Python commands. These commands can either come from the user or be generated by the LLM. This tool is particularly useful for complex calculations. Instead of having the LLM generate the answer directly, it can be more efficient to have the LLM generate code to calculate the answer.


In [30]:
from langchain.agents import Tool
from langchain_experimental.utilities import PythonREPL
python_repl = PythonREPL()
python_repl.run("a = 3; b = 1; print(a+b)")

/root/.local/lib/python3.11/site-packages/langchain_experimental/utilities/__init__.py:2: LangChainDeprecationWarning: As of langchain-core 0.3.0, LangChain uses pydantic v2 internally. The langchain.pydantic_v1 module was a compatibility shim for pydantic v1, and should no longer be used. Please update the code to import from Pydantic directly.

For example, replace imports like: `from langchain.pydantic_v1 import BaseModel`
with: `from pydantic import BaseModel`
or the v1 compatibility namespace if you are working in a code base that has not been fully upgraded to pydantic 2 yet. 	from pydantic.v1 import BaseModel

  from langchain_experimental.utilities.python import PythonREPL


'4\n'

##### Toolkits

Toolkits are collections of tools that are designed to be used together for specific tasks.

Let's create a toolkit that contains one tool which is `PythonREPLTool`. Note that tools are put into a `list` object.


In [31]:
from langchain_experimental.tools import PythonREPLTool
tools = [PythonREPLTool()]

/usr/local/lib/python3.11/dist-packages/pydantic/_internal/_generate_schema.py:913: UserWarning: Mixing V1 models and V2 models (or constructs, like `TypeAdapter`) is not supported. Please upgrade `PythonREPL` to V2.
  warn(


A list of toolkits that Langchain supports is available at [https://python.langchain.com/v0.1/docs/integrations/toolkits/](https://python.langchain.com/v0.1/docs/integrations/toolkits/).

By themselves, language models can't take actions - they just output text. A big use case for LangChain is creating agents. Agents are systems that use an LLM as a reasoning engineer to determine which actions to take and what the inputs to those actions should be. The results of those actions can then be fed back into the agent. The agent then makes a determination whether more actions are needed, or whether it is okay to finish.


In [33]:
from langchain.agents import create_react_agent
from langchain import hub
from langchain.agents import AgentExecutor

instructions = """You are an agent designed to write and execute python code to answer questions.
You have access to a python REPL, which you can use to execute python code.
If you get an error, debug your code and try again.
Only use the output of your code to answer the question.
You might know the answer without running any code, but you should still run the code to get the answer.
If it does not seem like you can write code to answer the question, just return "I don't know" as the answer.
"""

# here you will use the prompt directly from the langchain hub
base_prompt = hub.pull("langchain-ai/react-agent-template")
prompt = base_prompt.partial(instructions=instructions)

/root/.local/lib/python3.11/site-packages/langsmith/client.py:241: LangSmithMissingAPIKeyWarning: API key must be provided when using hosted LangSmith API
  warnings.warn(


You'll use the `create_react_agent` agent. It combines reasoning (e.g., Chain-of-Thought (CoT) prompting) and acting (e.g., action plan generation) together to let the LLM solve questions like humans would.

Now, set `verbose=True` to see how the LLM thinks and acts at every step.


In [35]:
agent = create_react_agent(llm, tools, prompt)
agent_executor = AgentExecutor(agent=agent, tools=tools, verbose=True, handle_parsing_errors=True)  # tools were defined in the toolkit part above

In [36]:
agent_executor.invoke(input = {"input": "What is the 3rd fibonacci number?"})

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.




> Entering new AgentExecutor chain...


/usr/local/lib/python3.11/dist-packages/transformers/generation/utils.py:2347: UserWarning: You are calling .generate() with the `input_ids` being on a device type different than your model's device. `input_ids` is on cpu, whereas the model is on cuda. You may experience unexpected behaviors or slower generation. Please make sure that you have put `input_ids` to the correct device by calling for example input_ids = input_ids.to('cuda') before running `.generate()`.
  warnings.warn(
Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.



New output: 3

New input: What is the 3rd fibonacci number?Invalid Format: Missing 'Action:' after 'Thought:'

/usr/local/lib/python3.11/dist-packages/transformers/generation/utils.py:2347: UserWarning: You are calling .generate() with the `input_ids` being on a device type different than your model's device. `input_ids` is on cpu, whereas the model is on cuda. You may experience unexpected behaviors or slower generation. Please make sure that you have put `input_ids` to the correct device by calling for example input_ids = input_ids.to('cuda') before running `.generate()`.
  warnings.warn(
Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.



New output: 3

New input: What is the 3rd fibonacci number?Invalid Format: Missing 'Action:' after 'Thought:'

/usr/local/lib/python3.11/dist-packages/transformers/generation/utils.py:2347: UserWarning: You are calling .generate() with the `input_ids` being on a device type different than your model's device. `input_ids` is on cpu, whereas the model is on cuda. You may experience unexpected behaviors or slower generation. Please make sure that you have put `input_ids` to the correct device by calling for example input_ids = input_ids.to('cuda') before running `.generate()`.
  warnings.warn(
Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.



New output: 3

New input: What is the 3rd fibonacci number?Invalid Format: Missing 'Action:' after 'Thought:'

/usr/local/lib/python3.11/dist-packages/transformers/generation/utils.py:2347: UserWarning: You are calling .generate() with the `input_ids` being on a device type different than your model's device. `input_ids` is on cpu, whereas the model is on cuda. You may experience unexpected behaviors or slower generation. Please make sure that you have put `input_ids` to the correct device by calling for example input_ids = input_ids.to('cuda') before running `.generate()`.
  warnings.warn(
Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.



New output: 3

New input: What is the 3rd fibonacci number?Invalid Format: Missing 'Action:' after 'Thought:'

/usr/local/lib/python3.11/dist-packages/transformers/generation/utils.py:2347: UserWarning: You are calling .generate() with the `input_ids` being on a device type different than your model's device. `input_ids` is on cpu, whereas the model is on cuda. You may experience unexpected behaviors or slower generation. Please make sure that you have put `input_ids` to the correct device by calling for example input_ids = input_ids.to('cuda') before running `.generate()`.
  warnings.warn(
Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.



New output: 3

New input: What is the 3rd fibonacci number?Invalid Format: Missing 'Action:' after 'Thought:'

/usr/local/lib/python3.11/dist-packages/transformers/generation/utils.py:2347: UserWarning: You are calling .generate() with the `input_ids` being on a device type different than your model's device. `input_ids` is on cpu, whereas the model is on cuda. You may experience unexpected behaviors or slower generation. Please make sure that you have put `input_ids` to the correct device by calling for example input_ids = input_ids.to('cuda') before running `.generate()`.
  warnings.warn(
Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.



New output: 3

New input: What is the 3rd fibonacci number?Invalid Format: Missing 'Action:' after 'Thought:'

/usr/local/lib/python3.11/dist-packages/transformers/generation/utils.py:2347: UserWarning: You are calling .generate() with the `input_ids` being on a device type different than your model's device. `input_ids` is on cpu, whereas the model is on cuda. You may experience unexpected behaviors or slower generation. Please make sure that you have put `input_ids` to the correct device by calling for example input_ids = input_ids.to('cuda') before running `.generate()`.
  warnings.warn(
Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.



New output: 3

New input: What is the 3rd fibonacci number?Invalid Format: Missing 'Action:' after 'Thought:'

/usr/local/lib/python3.11/dist-packages/transformers/generation/utils.py:2347: UserWarning: You are calling .generate() with the `input_ids` being on a device type different than your model's device. `input_ids` is on cpu, whereas the model is on cuda. You may experience unexpected behaviors or slower generation. Please make sure that you have put `input_ids` to the correct device by calling for example input_ids = input_ids.to('cuda') before running `.generate()`.
  warnings.warn(


KeyboardInterrupt: 